frontend/model/contracts/shared/functions.js

```diff
@@ -22,6 +23,62 @@ import { logExceptNavigationDuplicated } from '~/frontend/views/utils/misc.js'
 
 // group.js related
 
+export async function getPaymentsByPeriod (period: string) {
+  const rootGetters = sbp('state/vuex/getters')
+  const rootState = sbp('state/vuex/state')
+  const payments = rootGetters.paymentsForPeriod(period)
+  if (payments.length) {
+    return payments
+  }
+
+  // the rule to make key is there inside `archivePayments` function of group.js contract
+  const periodKey = `paymentPeriods/${rootGetters.ourUsername}/${rootState.currentGroupId}`
+  const periods = await sbp('gi.db/archive/load', periodKey) || []
+  if (periods.includes(period)) {
+    const paymentsKey = `paymentsByPeriod/${rootGetters.ourUsername}/${rootState.currentGroupId}/${period}`
+    const paymentsByHash = await sbp('gi.db/archive/load', paymentsKey) || {}
+    for (const hash of Object.keys(paymentsByHash)) {
+      const payment = paymentsByHash[hash]
+      if (payment.data.status === PAYMENT_COMPLETED) {
+        payments.push({
+          from: payment.meta.username,
+          to: payment.data.toUser,
+          hash,
+          amount: payment.data.amount,
+          isLate: !!payment.data.isLate,
+          when: payment.data.completedDate
+        })
```

Comment:
This code appears twice now, once here and once in the group contract getter `paymentsForPeriod`. We need to DRY this by creating a shared function that takes as input the payment and the payment hash, and call that function in both places. This will help prevent a divergence in the type of object returned.

Reply:
Nice catch.